Source/GSOrderedSet.m

```diff
@@ -0,0 +1,407 @@
+@private
+  NSUInteger _version;
+}
+@end
+
+@interface GSOrderedSetEnumerator : NSEnumerator
+{
+  GSOrderedSet *set;
+  unsigned      current;
+  unsigned      count;
+}
+@end
+
+@implementation GSOrderedSetEnumerator
+- (id) initWithOrderedSet: (NSOrderedSet*)d
+{
+  self = [super init];
+  if (self != nil)
+    {
+      set = (GSOrderedSet*)RETAIN(d);
+      current = 0;
+      count = GSIArrayCount(&set->array);
+    }
+  return self;
+}
+
+- (id) nextObject
+{
+  if(current < count)
+    {
+      GSIArrayItem item = GSIArrayItemAtIndex(&set->array, current);
+      current++;
+      return (id)(item.obj);
+    }
+  return nil;
+}
+
+- (void) dealloc
+{
+  RELEASE(set);
+  [super dealloc];
+}
+@end
+
+@implementation GSOrderedSet
+
+static Class	setClass;
+static Class	mutableSetClass;
+
++ (void) initialize
+{
+  if (self == [GSOrderedSet class])
+    {
+      setClass = [GSOrderedSet class];
+      mutableSetClass = [GSMutableOrderedSet class];
+      privateCountOfSel = @selector(_countForObject:);
+    }
+}
+
+- (id) copyWithZone: (NSZone*)z
+{
+  return RETAIN(self);
+}
+
+- (void) dealloc
+{
+  GSIArrayEmpty(&array);
+  [super dealloc];
+}
+
+- (NSUInteger) hash
+{
+  return [self count];
+}
+
+- (id) init
+{
+  return [self initWithObjects: 0 count: 0];
+}
+
+- (NSEnumerator*) objectEnumerator
+{
+  return AUTORELEASE([[GSOrderedSetEnumerator alloc] initWithOrderedSet: self]);
+}
+
+- (NSUInteger) countByEnumeratingWithState: (NSFastEnumerationState*)state
+                                   objects: (id*)stackbuf
+                                     count: (NSUInteger)len
+{
+  //state->mutationsPtr = (unsigned long *)self;
+  //return GSIMapCountByEnumeratingWithStateObjectsCount
+  //            (&map, state, stackbuf, len);
+
+  return [self count];
+}
+
+- (NSUInteger) sizeInBytesExcluding: (NSHashTable*)exclude
+{
+  NSUInteger	size = GSPrivateMemorySize(self, exclude);
+
+  if (size > 0)
+    {
+      NSUInteger count = [self count];
+      NSUInteger i = 0;
+
+      for(i = 0; i < count; i++)
+	{
+	  GSIArrayItem item = GSIArrayItemAtIndex(&array, i);
+          size += [item.obj sizeInBytesExcluding: exclude];
+        }
+    }
+  return size;
+}
+
+// Put required overrides here...
+- (NSUInteger) count
+{
+  return GSIArrayCount(&array);
+}
+
+- (id) objectAtIndex: (NSUInteger)index
+{
+  GSIArrayItem item = GSIArrayItemAtIndex(&array, index);
+  return item.obj;
+}
+
+- (id) objectAtIndexedSubscript: (NSUInteger)index
+{
+  return[self objectAtIndex: index];
+}
+
+- (BOOL) containsObject: (id)anObject
+{
+  NSUInteger i = 0;
+
+  for (i = 0; i < [self count]; i++)
+    {
+      id obj = [self objectAtIndex: i];
+      if([anObject isEqual: obj])
+	{
+	  return YES;
+	}
+    }
+  
+  return NO;
+}
+
+/* Designated initialiser */
+- (id) initWithObjects: (const id*)objs count: (NSUInteger)c
+{
+  NSUInteger i = 0;
+
+  // Reduce C if it doesn't match.  This is apparently how
+  // macOS behaves.
+  if(c > sizeof(objs))
```

Comment:
I am not sure whether this will work. How is the compiler able to find out about the real size of the array?

Reply:
Removed this